### PR TITLE
feat(mem): switch the default allocator to tlsf

### LINKED ISSUE #2123: Replace LVGL's memory allocator with TLSF
As discussed in #2116 LVGL's built-in memory manager could be improved in terms of performance and fragmentation. [tlsf](https://github.com/mattconte/tlsf) seems like a good candidate.

At first glance, the built-in current memory allocator could be replaced by tlsf without any API change. Even the memory monitor and memory test features can be kept.

This list summarizes requirements:
- The following top level functions are required: 
  - `lv_mem_alloc` (rename to `lv_malloc`?)
  - `lv_mem_free` (rename to `lv_free`?)
  - `lv_mem_realloc` (rename to `lv_realloc`?)
- These functions are also wrappers to `LV_MEM_CUSTOM_ALLOC/FREE/REALLOC` (set in `lv_conf.h` if `LV_MEM_CUSTOM 1`)
- The size of the memory pool should be configurable via `LV_MEM_SIZE`. The max size can be very large too (even hundreds of MB-s). If possible optimize tlsf for certain ranges via `#if`s. E.g. 0..8kB, 8kb..32kB, 32 kb...256kB, >1MB or so.  
- Manual defrag (now `lv_mem_defrag`) is not required as it should be handled in tlsf.
- Allow allocating 0 sized memory. Now a dedicated pointer is used in this case (called `zero_mem`).
- Reimplement `lv_mem_monitor` to extract at least the free and used size, largest free block and calculate the fragmentation
- Reimplement `lv_mem_test` to check memory integrity.
- Other functions in `lv_mem.c` shouldn't be affected (E.g. `lv_mem_buf_...` , `lv_memcpy`, `lv_memset`, etc)
- `tlsf.c/h` should be added directly as `lv_tlsf` (?) with correct attribution instead of submodules because submodules might be problematic for some package managers or for novice users. This way `lv_mem.c` can be clean containing only tlsf or `LV_MEM_CUSTOM` calls.

@xiaoxiang781216 already has experience with tlsf and mentioned that he interested in adding tlsf. Thank your for this! :slightly_smiling_face:

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
### Description of the feature or fix

See https://github.com/lvgl/lvgl/issues/2123 for more info.

### Checkpoints
- [ ] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Update CHANGELOG.md
- [ ] Update the documentation

### CODE REVIEW

**Comment by puzrin** [comment]
Is that magick with `mem_buf_small...` still required? Looks like outdated micro-optimisation for couple of bytes. 

@kisvegabor what cases it reflects?

**Comment by xiaoxiang781216** [comment]
Yes, `mem_buf_small...` could be removed since tlsf is enough fast to make the micro-optimisation lose the meaning.

**Comment by embeddedt** [comment]
The only issue I see is that a project using TLSF already will start getting link errors. Perhaps it's a good idea to add `#if LV_MEM_CUSTOM == 0` to the `lv_tlsf.h` and `lv_tlsf.c` files so it can be disabled fully?

**Comment by xiaoxiang781216** [comment]
> The only issue I see is that a project using TLSF already will start getting link errors. Perhaps it's a good idea to add `#if LV_MEM_CUSTOM == 0` to the `lv_tlsf.h` and `lv_tlsf.c` files so it can be disabled fully?

Done.

**Comment by puzrin** [comment]
Not allocator question, but nearby... in scope of "let's remove all hand-mades"

Are lv_memcpy/lv_memset required? Default ones from C are not optimised for 32 bits? Or any compiler has problems with those?

/cc @kisvegabor

**Comment by embeddedt** [comment]
They were added in v7 because we found that ours are often faster than the ones that platforms provide by default.

They can be disabled [in `lv_conf.h`](https://github.com/lvgl/lvgl/blob/607e02e52ae4201a8f9830a078be17abc01a3cad/lv_conf_template.h#L105).

**Comment by puzrin** [comment]
> They were added in v7 because we found that ours are often faster than the ones that platforms provide by default.

IMHO it would be nice to add code comments about platforms/compilers where such manual unroll make sense. Because that's not obvious (and doing compiler's job by hands is not nice, IMO). If you wish, i can create new issue later, to avoid buzz in this PR.

PS. Probably, i'm not right - not enough experience with c/cpp to insist.

**Comment by kisvegabor** [comment]
Looks awesome, thank you!

I've tested it and got two questions:
1. Would it be possible to minimize the memory usage of tlsf by reconfiguring it based on `LV_MEM_SIZE`? 
2. I've tested it with a UI that uses 16kB by default with the old allocator. Now it's 24kB. The memory required by tlsf is counted into `LV_MEM_SIZE`?

**Comment by xiaoxiang781216** [comment]
> Looks awesome, thank you!
> 
> I've tested it and got two questions:
> 
> 1. Would it be possible to minimize the memory usage of tlsf by reconfiguring it based on `LV_MEM_SIZE`?

Yes, we can tune the table base on `LV_MEM_SIZE`, I will update the patch.

> 2. I've tested it with a UI that uses 16kB by default with the old allocator. Now it's 24kB. The memory required by tlsf is counted into `LV_MEM_SIZE`?

Yes, TLFS reserved some space from `LV_MEM_SIZE` and don't define or use any global variables.

**Comment by xiaoxiang781216** [comment]
> Looks awesome, thank you!
> 
> I've tested it and got two questions:
> 
> 1. Would it be possible to minimize the memory usage of tlsf by reconfiguring it based on `LV_MEM_SIZE`?

Done, the inital overhead is reduced from ~3KB to ~1KB with the default config(LV_MEM_SIZE=32KB).
Note: the value is based on 32bit platform, will double on 64bit platform.

**Comment by puzrin** [comment]
> the inital overhead is reduced from ~3KB to ~1KB with the default config(LV_MEM_SIZE=32KB).

Nice. Could you say overhead for LV_MEM_SIZE of 5К-8К?

**Comment by xiaoxiang781216** [comment]
> > the inital overhead is reduced from ~3KB to ~1KB with the default config(LV_MEM_SIZE=32KB).
> 
> Nice. Could you say overhead for LV_MEM_SIZE of 5К-8К?

Here is the structure used by tlsf to manage the heap:
```
/* The TLSF control structure. */
typedef struct control_t
{
	/* Empty lists point at this block to indicate they are free. */
	block_header_t block_null;------------------------------------------>   16(   32)

	/* Bitmaps for free lists. */
	unsigned int fl_bitmap;--------------------------------------------->    4(    4)
	unsigned int sl_bitmap[FL_INDEX_COUNT];----------------------------->  4*n(  4*n)

	/* Head of free lists. */
	block_header_t* blocks[FL_INDEX_COUNT][SL_INDEX_COUNT];------------->4*n*m(8*n*m)
} control_t;
```
m(SL_INDEX_COUNT) could be 8, 16, and 32, but the default is 32 to get the smallest internal fragmentation.
Here is the equation for n(FL_INDEX_COUNT): 
```
FL_INDEX_COUNT      = (FL_INDEX_MAX - FL_INDEX_SHIFT + 1)
FL_INDEX_SHIFT      = (SL_INDEX_COUNT_LOG2 + ALIGN_SIZE_LOG2)
SL_INDEX_COUNT_LOG2 = log2(SL_INDEX_COUNT) = 5
ALIGN_SIZE_LOG2     = 2(3)
FL_INDEX_MAX        = log2(size)
```
and then n=log2(size)-6(7) and sizeof(control_t) equal:
```
 = 16 + 4 + 4*n + 4*n*m(32 + 4 + 4*n + 8*n*m)
 = 132*n + 20(250*n + 36)
 = 132*log2(size) - 772(250*log2(size) - 1714)
```
so here is a table for the typical size:
```
 8KB |  944(1640)
16KB | 1076(1786)
32KB | 1208(2036)
 1GB | 3188(5786)
 2GB | 3320(6036)
```

**Comment by puzrin** [comment]
Thanks for details. So, we can say, for stm32f0/f4 quick-estimated overhead will be 1K. Looks good.

**Comment by kisvegabor** [comment]
@xiaoxiang781216 I've tested it and now the memory usage is quite the same as with the original allocator. 

The stress test is running well too:
stress

Only one thing: in `lv_mem_walker` doesn't we need to count the 4 byte block header too?

File: src/lv_misc/lv_tlsf.c [hunk 1]
```diff
@@ -0,0 +1,1286 @@
...
+
+	return status;
+}
+
+#undef tlsf_insist
+
+static void default_walker(void* ptr, size_t size, int used, void* user)
+{
+	(void)user;
+	printf("\t%p %s size: %x (%p)\n", ptr, used ? "used" : "free", (unsigned int)size, (void*)block_from_ptr(ptr));
```
Comment (kisvegabor) [inline src/lv_misc/lv_tlsf.c:925]: This `printf` should be removed (or the whole function) because it might cause an unexpected bloat in the code size.

**Comment by kisvegabor** [comment]
That's true, okay.

File: src/lv_misc/lv_tlsf.c (hunk 1 above)
Comment (xiaoxiang781216) [reply src/lv_misc/lv_tlsf.c:925]: I add a defintion at the beginning:
```
#include "lv_log.h"
#define printf LV_LOG_ERROR
```
to redirect printf to lvgl log. It should be safe now.

Comment (kisvegabor) [reply src/lv_misc/lv_tlsf.c:925]: Perfect, thanks!

**Comment by kisvegabor** [comment]
Is this PR ready to merge?

**Comment by xiaoxiang781216** [comment]
Yes, I think so.

**Review by embeddedt — APPROVED** [review]

**Pull request merged**